README.md
# utest: Asynchronous C++ Test Harness

This test harness allows you to execute a specified series of (asynchronous) C++ test cases with sensible default reporting and useful customization options.

Please note that this is a purposefully lean test harness, only dealing with test execution and providing default reporting handlers. It specifically does not support auto-discovery of test cases and does not provide you with test macros or other convenience functions.
Instead, the macros in the [unity module](https://github.com/ARMmbed/unity) can be used for this purpose. However, you are not required to use these, and can use your own macros if you wish.

Furthermore, test failure recovery through the use of exceptions or `longjmp` is not supported; the test will either continue and ignore failures or die by busy-waiting.

## Theory of Operation

A test specification contains a setup handler, several test cases and a teardown handler.
Each test case contains a textual description, setup, teardown and failure handler as well as the actual test handler.

The order of handler execution is:

1. Test setup handler.
1. For each test case:
    1. Test case setup handler.
    1. Test case execution handler.
    1. (wait for timeout or callback validation in case of an asynchronous test case.)
    1. (repeat test case execution handler if specified.)
    1. Test case teardown handler.
1. Test teardown handler.

## Example

The following example showcases a lot of functionality and proper integration with the [Greentea testing automation framework](https://github.com/ARMmbed/greentea), while making use of the [unity test macros](https://github.com/ARMmbed/unity):

```cpp
#include "mbed-drivers/test_env.h"
#include "utest/utest.h"
#include "unity/unity.h"

using namespace utest::v1;

void test_simple() {
    TEST_ASSERT_EQUAL(0, 0);
    printf("Simple test called\n");
}

status_t test_repeats_setup(const Case *const source, const size_t index_of_case) {
    // Call the default handler for proper reporting
    status_t status = greentea_case_setup_handler(source, index_of_case);
    printf("Setting up for '%s'\n", source->get_description());
    return status;
}
control_t test_repeats(const size_t call_count) {
    printf("Called for the %u. time\n", call_count);
    TEST_ASSERT_NOT_EQUAL(3, call_count);
    // Specify how often this test is repeated ie. n total calls
    return (call_count < 2) ? CaseRepeatAll : CaseNext;
}

void test_callback_validate() {
    // You may also use assertions here!
    TEST_ASSERT_EQUAL_PTR(0, 0);
    // Validate the callback
    Harness::validate_callback();
}
control_t test_asynchronous() {
    TEST_ASSERT_TRUE_MESSAGE(true, "(true == false) o_O");
    // Set up a callback in the future. This may also be an interrupt!
    minar::Scheduler::postCallback(test_callback_validate).delay(minar::milliseconds(100));
    // Set a 200ms timeout starting from now
    return CaseTimeout(200);
}

control_t test_asynchronous_timeout(const size_t call_count) {
    TEST_ASSERT_TRUE_MESSAGE(true, "(true == false) o_O");
    // Set a 200ms timeout starting from now,
    // but automatically repeat only this handler on timeout.
    if (call_count >= 5) {
        // but after the 5th call, the callback finally gets validated
        minar::Scheduler::postCallback(test_callback_validate).delay(minar::milliseconds(100));
    }
    return CaseRepeatHandlerOnTimeout(200);
}

// Custom setup handler required for proper Greentea support
status_t greentea_setup(const size_t number_of_cases) {
    GREENTEA_SETUP(20, "default_auto");
    // Call the default reporting function
    return greentea_test_setup_handler(number_of_cases);
}

// Specify all your test cases here
Case cases[] = {
    Case("Simple Test", test_simple),
    Case("Repeating Test", test_repeats_setup, test_repeats),
    Case("Asynchronous Test (200ms timeout)", test_asynchronous),
    Case("Asynchronous Timeout Repeat", test_asynchronous_timeout)
};

// Declare your test specification with a custom setup handler
Specification specification(greentea_setup, cases);

void app_start(int, char**)
{   // Run the test specification
    Harness::run(specification);
}
```

Running this test will output the following:

```
{{timeout;20}}
{{host_test_name;default_auto}}
{{description;utest greentea example}}
{{test_id;MBED_OS}}
{{start}}
>>> Running 4 test cases...

>>> Running case #1: 'Simple Test'...
Simple test called
>>> 'Simple Test': 1 passed, 0 failed

>>> Running case #2: 'Repeating Test'...
Setting up for 'Repeating Test'
Called for the 1. time
>>> 'Repeating Test': 1 passed, 0 failed

>>> Running case #2: 'Repeating Test'...
Setting up for 'Repeating Test'
Called for the 2. time
>>> 'Repeating Test': 2 passed, 0 failed

>>> Running case #3: 'Asynchronous Test (200ms timeout)'...
>>> 'Asynchronous Test (200ms timeout)': 1 passed, 0 failed

>>> Running case #4: 'Asynchronous Timeout Repeat'...
>>> failure with reason 'Ignored: Timed Out'
>>> failure with reason 'Ignored: Timed Out'
>>> failure with reason 'Ignored: Timed Out'
>>> failure with reason 'Ignored: Timed Out'
>>> failure with reason 'Ignored: Timed Out'
>>> 'Asynchronous Timeout Repeat': 1 passed, 0 failed

>>> Test cases: 4 passed, 0 failed
{{success}}
{{end}}
```

## Detailed Description

### Handlers

There are five handler types you can, but do not have to, override to customize operation.
Please see the `utest/types.h` file for a detailed description.

1. `status_t test_setup_handler_t(const size_t number_of_cases)`: called before execution of any test case.
1. `void test_teardown_handler_t(const size_t passed, const size_t failed, const failure_t failure)`: called after execution of all test cases, and if testing is aborted.
1. `void test_failure_handler_t(const failure_t failure)`: called whenever a failure occurs anywhere in the specification.
1. `status_t case_setup_handler_t(const Case *const source, const size_t index_of_case)`: called before execution of each test case.
1. `status_t case_teardown_handler_t(const Case *const source, const size_t passed, const size_t failed, const failure_t reason)`: called after execution of each test case, and if testing is aborted.
1. `status_t case_failure_handler_t(const Case *const source, const failure_t reason)`: called whenever a failure occurs during the execution of a test case.

All handlers are defaulted for integration with the [Greentea testing automation framework](https://github.com/ARMmbed/greentea).

### Test Case Handlers

There are three test case handlers:

1. `void case_handler_t(void)`: executes once, if the case setup succeeded.
1. `control_t case_control_handler_t(void)`: executes (asynchronously) as many times as you specify, if the case setup succeeded.
1. `control_t case_call_count_handler_t(const size_t call_count)`: executes (asynchronously) as many times as you specify, if the case setup succeeded.

To specify a test case you must wrap it into a `Case` class: `Case("mandatory description", case_handler)`. You may override the setup, teardown and failure handlers in this wrapper class as well.
The `Case` constructor is overloaded to allow you a comfortable declaration of all your callbacks and the order of arguments is:

1. Description (required).
1. Setup handler (optional).
1. Test case handler (required).
1. Teardown handler (optional).
1. Failure handler (optional).

#### Test Case Attributes

You can modify test case behavior by returning `control_t` modifiers:

- `CaseNext`: never repeats and immediately moves to next test case
- `CaseNoRepeat`: never repeats.
- `CaseRepeatAll`: repeats test case **with** setup and teardown handlers.
- `CaseRepeatHandler`: repeats test case **without** set and teardown handlers.
- `CaseNoTimeout`: immediately moves to next test case.
- `CaseAwait`: waits indefinitely for callback validation (*use with caution*).
- `CaseTimeout(uint32_t ms)`: waits for callback validation for `ms` milliseconds, times out after that (fails with `REASON_TIMEOUT`).
- `CaseRepeatAllOnTimeout(uint32_t ms)`: waits for callback validation for `ms` milliseconds, repeats test case **with** setup and teardown handlers on time out.
- `CaseRepeatHandlerOnTimeout(uint32_t ms)`: waits for callback validation for `ms` milliseconds, repeats test case **without** setup and teardown handlers on time out.

Returning `CaseRepeatAll` from your test case handler tells the test harness to repeat the test handler. You can use the `call_count` (starts counting at 1) to decide when to stop.
By default the setup and teardown handlers are called on every repeated test cases, however, you may only repeat the case handler by returning `CaseRepeatHandler`. To stop the harness from repeating the test case, return `CaseNext`.

For asynchronous test cases, you must return a `CaseTimeout(uint32_t ms)`.
If you want to automatically repeat the test case on a timeout, use `CaseRepeatAllOnTimeout(uint32_t ms)` and `CaseRepeatHandlerOnTimeout(uint32_t ms)`.

To validate your callback, you must call `Harness::validate_callback()` in your asynchronous callback before the timeout fires.
This will schedule the execution of the next test case.

For repeating asynchronous cases, you can "add" both attributes together: `CaseTimeout(200) + CaseRepeatAll` will wait for 200ms for the callback validation and then repeat the test case. See the section on arbitration logic for more details.

Note that you can also add attributes during callback validation, however, only repeat attributes are considered. This allows you to return `CaseTimeout(500)` to wait up to 500ms for the callback validation and delegate the decision to repeat to the time the callback occurs: `Harness::validate_callback(CaseRepeatHandler)`.

Keep in mind that you can only validate a callback once. If you need to wait for several callbacks, you need to write your own helper function that validates the expected callback only when all your custom callbacks arrive.
This custom functionality is purposefully not part of this test harness, you can achieve it externally with additional code.

### Failure Handlers

A failure may occur during any phase of the test. The appropriate failure handler is then called with `failure_t`, which contains the failure reason and location.

The failure reasons are:

- `REASON_NONE`: No failure occurred
- `REASON_UNKNOWN`: An unknown failure occurred
- `REASON_CASES`: A failure occurred in at least one test case
- `REASON_EMPTY_CASE`: The test case contains only empty handlers
- `REASON_TIMEOUT`: An expected asynchronous call timed out
- `REASON_ASSERTION`: An assertion failed
- `REASON_TEST_SETUP`: Test setup failed
- `REASON_TEST_TEARDOWN`: Test teardown failed
- `REASON_CASE_SETUP`: Case setup failed
- `REASON_CASE_HANDLER`: Case handler failed
- `REASON_CASE_TEARDOWN`: Case teardown failed
- `REASON_CASE_INDEX`: Case index returned from test setup or case teardown handler is invalid
- `REASON_SCHEDULER`: Underlying scheduler is not asynchronous

The failure locations are:

- `LOCATION_NONE`: No location information
- `LOCATION_UNKNOWN`: A failure occurred in an unknown location
- `LOCATION_TEST_SETUP`: A failure occurred in the test setup
- `LOCATION_TEST_TEARDOWN`: A failure occurred in the test teardown
- `LOCATION_CASE_SETUP`: A failure occurred in the case setup
- `LOCATION_CASE_HANDLER`: A failure occurred in the case handler
- `LOCATION_CASE_TEARDOWN`: A failure occurred in the case teardown

If the setup or teardown handlers fail, they may return a `STATUS_ABORT` code, which will call the failure handler with the appropriate failure reason (`REASON_CASE_{SETUP|TEARDOWN}`) and failure location (`LOCATION_CASE_{SETUP|TEARDOWN}`).
If the setup handler fails, the test case is never executed. Instead, the teardown handler is called in an attempt to salvage the situation.
Please note that if a teardown handler fails, the system can be considered too unstable to continue testing.

You may also raise a failure manually by calling `Harness::raise_failure(failure_reason_t reason)`. In fact, this is how you can integrate assertion failures from custom test macros, as done with the unity macros, which raise a failure with the `REASON_ASSERTION` reason.

When waiting for an asynchronous callback, if the timeout fires, `REASON_TIMEOUT` is raised.

The failure handler decides whether to continue or abort testing by returning `STATUS_CONTINUE` or `STATUS_ABORT` respectively.
You can also ignore any raised failure by returning `STATUS_IGNORE` and the harness will then not count this failure.
In case of an abort, the test harness dies by busy waiting in a forever loop.
This is needed because we cannot unwind the stack without exception support, and the asynchronous nature of the test harness breaks with using `longjmp`s.

Note that when `REASON_IGNORE` is `OR`ed into the failure reason, the failure handler is expected to return `STATUS_IGNORE`.
This is done automatically for test cases repeating after a timeout, and the default failure handlers also report this failure, but tell the harness to ignore it.
Furthermore, the unity macros may decide to ignore assertion failures as well, in which case the assertion is ignored intentionally.

### Default Handlers

Three sets of default handlers with different behaviors are provided for your convenience:

1. `greentea_abort_handlers` (default): Greentea-style reporting, aborts on the first failure, but requires custom test setup handler.
1. `greentea_continue_handlers`: Greentea-style reporting, always continues testing, but requires custom test setup handler.
1. `verbose_continue_handlers`: always continues testing and reporting, except when a teardown failed.
1. `selftest_handlers`: Greentea-style reporting, but aborts on the first assertion failure raised. This allows the use of unity macros for self testing without recursive failure handler calls.

These default handlers are called when you have not overridden a custom handler, and they only contain reporting functionality and do not modify global state.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
## [1.12.2] - 2016-03-31
### Added
- Also `exit(1)` on test failure.

## [1.12.1] - 2016-03-30
### Added
- `exit(test_result)` after harness finished execution.

## [1.12.0] - 2016-03-23
- Release of mbed OS 16.03.

## [1.11.0] - 2016-03-16
### Added
- Updated dependency on greentea-client.

## [1.10.1] - 2016-03-15
### Added
- Coverage configuration file.

### Fixed
- Removed duplicate dependency.

## [1.10.0] - 2016-03-02
### Added
- Support for `greentea-client` APIs: preamble, hand-shake and test suite setup.
- Support for `Greentea` automation tool: Additional KiVi (key-value) prints.
- Separate Greentea handlers section.
- Excluded lambdas (C++11) code failing to compile on ARMCCv5.

## [1.9.2] - 2016-02-29
### Fixed
- Calling `raise_failure` before harness is initialized won't have an effect.

## [1.9.1] - 2016-01-25
### Fixed
- Also check `control_handler` in `Case::is_empty()`.

## [1.9.0] - 2016-01-07
### Added
- Location information to `failure_t` and report it in failure handlers.
- Added documentation about failure reasons and locations to Readme.

### Fixed
- Assertion failures on test setup and test teardown will now not be ignored.

## [1.8.2] - 2015-12-23
### Added
- Code examples from Readme as tests.

### Changed
- Updated Readme with more documentation.

## [1.8.1] - 2015-12-22
### Added
- A few more test cases for case repeat counting.
- Self test for case timeout handling.
- Self test for case teardown failure.
- Self test for callback validation (with attributes).

### Changed
- Rename `case_control_handler` test to `case_control_repeat`.

### Fixed
- Modifying case attributes during callback validation.

## [1.8.0] - 2015-12-21
### Added
- Self test for case setup failure.
- Self test for case handler repeat counting.
- Explicit constants of each handler type to `{ignore|default}_handler`.
- Case attribute modification in `validate_callback`.

### Fixed
- Call count in example in Readme.

### Changed
- Replace `repeat_count` with `call_count`, which starts at 1 not at 0.

## [1.7.4] - 2015-12-17
### Added
- Added `CaseNoRepeat` and `CaseNoTimeout` aliases.

### Fixed
- Arbitration behavior when adding `CaseNoTimeout` to `CaseRepeat{All|Handler}OnTimeout`.

## [1.7.3] - 2015-12-16
### Fixed
- Order of unity macro arguments. They were backwards before.

## [1.7.2] - 2015-12-16
### Added
- Unit test for asserting `control_t` arbitration behavior.

### Fixed
- Clarify and implement `control_t` arbitration behavior.

## [1.7.1] - 2015-12-15
### Added
- This changelog  }:-o

## [1.7.0] - 2015-12-15
### Added
- `CaseAwait` control modifier and logic: Makes a case wait without timeout until a callback is validated.
- Explicit `TIMEOUT_*` enumeration for the magic values.

### Changed
- `CaseTimeout` and `CaseRepeat{All|Handler}OnTimeout` modifiers are now constructed using factory functions, rather than inheritance. This allows the compiler to infer the correct return type for the ternary operator.

## [1.6.0] - 2015-12-11
### Added
- Missing implementations for all callback handlers for all (especially the greentea) behaviors. These implementations only forward to the verbose handlers, however, should we modify the reporting style in the future, tests using the behavior-specific handlers will not have to be adapted.

### Changed
- Updated examples to call the behavior-specific handlers for reporting.

## [1.5.1] - 2015-12-10
### Added
- Self test for setup failure.

### Fixed
- Bugs in repeat logic.
- Bugs in index counting.
- Allow repeat on timeout and normal repeat at the same time.

## [1.5.0] - 2015-12-10
### Added
- Callback hook for unity's ignore handler.

## [1.4.0] - 2015-12-10
### Added
- `CaseRepeat{All|Handler}OnTimeout`: allows to repeat the test case automatically on timeout.
- Stringify for `status_t`.
- Allow ignoring of any failures using modifier mask.
- Global specification failure handler, which is called anytime a failure occurs, even on setup and teardown of specifications. This also allows for self testing using unity macros, as it prevents recursive case failure handler calling (think of a failed assert in the case failure handler).

### Changed
- deprecate `CaseRepeat` for `CaseRepeatAll` and `CaseRepeatHandlerOnly` for `CaseRepeatHandler`.

## [1.3.0] - 2015-12-08
### Changed
- Do not need to execute handlers atomically anymore. This interfered with test cases that required polling of interrupts.

## [1.2.0] - 2015-12-08
### Changed
- `Harness::run` returns whether the specification is valid and is being run now
- After specification teardown harness does not die anymore.

## [1.1.0] - 2015-12-04
### Changed
- `Harness::run` allows starting from a different test case than zero. This can be used

## [1.0.1] - 2015-11-25
### Changed
- Updated unity module test dependency to v2.0.0

## 1.0.0 - 2015-11-25
### Added
- v1 of this small test harness. See Readme.

[Unreleased]: https://github.com/ARMmbed/utest/compare/v1.12.2...HEAD
[1.12.2]: https://github.com/ARMmbed/utest/compare/v1.12.1...v1.12.2
[1.12.1]: https://github.com/ARMmbed/utest/compare/v1.12.0...v1.12.1
[1.12.0]: https://github.com/ARMmbed/utest/compare/v1.11.0...v1.12.0
[1.11.0]: https://github.com/ARMmbed/utest/compare/v1.10.1...v1.11.0
[1.10.1]: https://github.com/ARMmbed/utest/compare/v1.10.0...v1.10.1
[1.10.1]: https://github.com/ARMmbed/utest/compare/v1.10.0...v1.10.1
[1.10.0]: https://github.com/ARMmbed/utest/compare/v1.9.2...v1.10.0
[1.9.2]: https://github.com/ARMmbed/utest/compare/v1.9.1...v1.9.2
[1.9.1]: https://github.com/ARMmbed/utest/compare/v1.9.0...v1.9.1
[1.9.0]: https://github.com/ARMmbed/utest/compare/v1.8.2...v1.9.0
[1.8.2]: https://github.com/ARMmbed/utest/compare/v1.8.1...v1.8.2
[1.8.1]: https://github.com/ARMmbed/utest/compare/v1.8.0...v1.8.1

utest/types.h
/****************************************************************************
 * Copyright (c) 2015, ARM Limited, All Rights Reserved
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ****************************************************************************
 */

#ifndef UTEST_TYPES_H
#define UTEST_TYPES_H

#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include "compiler-polyfill/attributes.h"

namespace utest {
namespace v1 {

    enum repeat_t {
        REPEAT_UNDECLR = 0,
        REPEAT_NONE    = 1,       ///< continue with the next test case

        REPEAT_ON_TIMEOUT     =  2,
        REPEAT_ON_VALIDATE    =  4,
        REPEAT_CASE_ONLY      =  8,
        REPEAT_SETUP_TEARDOWN = 16,

        REPEAT_MASK = REPEAT_ON_TIMEOUT | REPEAT_ON_VALIDATE | REPEAT_CASE_ONLY | REPEAT_SETUP_TEARDOWN,

        REPEAT_ALL_ON_TIMEOUT = REPEAT_SETUP_TEARDOWN | REPEAT_ON_TIMEOUT,  ///< repeat the handler with setup and teardown on timeout
        REPEAT_HANDLER_ON_TIMEOUT = REPEAT_CASE_ONLY | REPEAT_ON_TIMEOUT,   ///< repeat only the handler on timeout
        REPEAT_ALL = REPEAT_SETUP_TEARDOWN | REPEAT_ON_VALIDATE,            ///< repeat the handler with setup and teardown
        REPEAT_HANDLER = REPEAT_CASE_ONLY | REPEAT_ON_VALIDATE              ///< repeat only the handler
    };

    enum status_t {
        STATUS_CONTINUE = 0,    ///< continues testing
        STATUS_IGNORE = -1,     ///< ignores failure and continues testing
        STATUS_ABORT = -2       ///< stops testing
    };

    enum failure_reason_t {
        REASON_NONE          = 0,           ///< No failure occurred

        REASON_UNKNOWN       = (1 << 0),    ///< An unknown failure occurred
        REASON_CASES         = (1 << 1),    ///< A failure occurred in at least one test case
        REASON_EMPTY_CASE    = (1 << 2),    ///< The test case contains only empty handlers
        REASON_TIMEOUT       = (1 << 3),    ///< An expected asynchronous call timed out
        REASON_ASSERTION     = (1 << 4),    ///< An assertion failed

        REASON_TEST_SETUP    = (1 << 5),    ///< Test setup failed
        REASON_TEST_TEARDOWN = (1 << 6),    ///< Test teardown failed
        REASON_CASE_SETUP    = (1 << 7),    ///< Case setup failed
        REASON_CASE_HANDLER  = (1 << 8),    ///< Case handler failed
        REASON_CASE_TEARDOWN = (1 << 9),    ///< Case teardown failed

        REASON_START_CASE    = (1 << 10),   ///< Start case index out-of-range

        REASON_IGNORE        = 0x8000       ///< The failure may be ignored
    };

    enum location_t {
        LOCATION_NONE = 0,      ///< No location information
        LOCATION_TEST_SETUP,    ///< A failure occurred in the test setup
        LOCATION_TEST_TEARDOWN, ///< A failure occurred in the test teardown
        LOCATION_CASE_SETUP,    ///< A failure occurred in the case setup
        LOCATION_CASE_HANDLER,  ///< A failure occurred in the case handler
        LOCATION_CASE_TEARDOWN, ///< A failure occurred in the case teardown
        LOCATION_UNKNOWN        ///< A failure occurred in an unknown location
    };

    /// Contains the reason and location of the failure.
    struct failure_t {
        failure_t(failure_reason_t reason) : reason(reason) {}
        failure_t(location_t location) : location(location) {}
        failure_t(failure_reason_t reason, location_t location) : reason(reason), location(location) {}

        /// @returns a copy of the failure with the reason ignored.
        failure_t ignored() const {
            return failure_t(failure_reason_t(reason | REASON_IGNORE), location);
        }

        failure_reason_t reason = REASON_NONE;
        location_t location = LOCATION_NONE;
    };


    enum {
        TIMEOUT_NONE    = uint32_t(-1), ///< Do not use a timeout
        TIMEOUT_UNDECLR = uint32_t(-2), ///< Timeout not explicitly specified, defaults to NONE
        TIMEOUT_FOREVER = uint32_t(-3)  ///< Never time out
    };

    /// Stringifies a failure reason for understandable error messages.
    const char* stringify(failure_reason_t reason);
    /// Stringifies a failure for understandable error messages.
    const char* stringify(failure_t failure);
    /// Stringifies a location.
    const char* stringify(location_t location);
    /// Stringifies a status.
    const char* stringify(status_t status);

    /** Control class for specifying test case attributes
     *
     * This class encapsulated control information about test cases which, when returned from
     * a test case influences the behavior of the test harness.
     * Instead of using this class directly it is recommended to use the aliases for clearer
     * semantics:
     * @code
     * control_t test_case(const size_t call_count) {
     *     // repeat 5 times for a total of 6 calls
     *     return (call_count < 6) ? CaseRepeatHandler : CaseNext;
     * }
     * @endcode
     *
     * This class overloads the `+` operator to implement something similiar to saturated arbitration:
     * - The lower timeout value "wins".
     * - A more involved repeat "wins" (ie. `ALL` > 'HANDLER' > 'NONE').
     * - Next Case always wins.
     *
     * You may then add timeouts and repeats together:
     * @code
     * control_t test_case(const size_t call_count) {
     *     // repeat 5 times for a total of 6 calls, each with a 500ms asynchronous timeout
     *     return CaseTimeout(500) + ((call_count < 6) ? CaseRepeatAll : CaseNoRepeat);
     * }
     * @endcode
     *
     * In the future, more control information may be added transparently and backwards compatible.
     */
    struct control_t
    {
        control_t() : repeat(REPEAT_UNDECLR), timeout(TIMEOUT_UNDECLR) {}

        control_t(repeat_t repeat, uint32_t timeout_ms) :
            repeat(repeat), timeout(timeout_ms) {}

        control_t(repeat_t repeat) :
            repeat(repeat), timeout(TIMEOUT_UNDECLR) {}

        control_t(uint32_t timeout_ms) :
            repeat(REPEAT_UNDECLR), timeout(timeout_ms) {}

        control_t
        inline operator+(const control_t& rhs) const {
            control_t result(
                repeat_t(this->repeat | rhs.repeat),
                (rhs.timeout == TIMEOUT_NONE) ? rhs.timeout : this->timeout);

            if (result.timeout != TIMEOUT_NONE && result.timeout > rhs.timeout) {
                result.timeout = rhs.timeout;
            }

            if (result.repeat & REPEAT_NONE) {
                result.repeat = REPEAT_NONE;
            }
            else {
                if (result.repeat & REPEAT_SETUP_TEARDOWN) {
                    result.repeat = repeat_t(result.repeat & ~REPEAT_CASE_ONLY);
                }
                if (result.timeout == TIMEOUT_NONE && result.repeat & REPEAT_ON_TIMEOUT) {
                    result.repeat = repeat_t(result.repeat & ~REPEAT_ON_TIMEOUT);
                }
            }

            return result;
        }

        repeat_t
        inline get_repeat() const {
            return repeat;
        }
        uint32_t
        inline get_timeout() const {
            return timeout;
        }

    private:
        repeat_t repeat;
        uint32_t timeout;
        friend class Harness;
    };

    /// does not repeat this test case and immediately moves on to the next one without timeout
    const  control_t CaseNext(REPEAT_NONE, TIMEOUT_NONE);

    /// does not repeat this test case, moves on to the next one
    const  control_t CaseNoRepeat(REPEAT_NONE);
    /// repeats the test case handler with calling teardown and setup handlers
    const  control_t CaseRepeatAll(REPEAT_ALL);
    /// repeats only the test case handler without calling teardown and setup handlers
    const  control_t CaseRepeatHandler(REPEAT_HANDLER);

    /// No timeout, immediately moves on to the next case, but allows repeats
    const  control_t CaseNoTimeout(TIMEOUT_NONE);
    /// Awaits until the callback is validated and never times out. Use with caution!
    const  control_t CaseAwait(TIMEOUT_FOREVER);
    /// Alias class for asynchronous timeout control in milliseconds
    inline control_t CaseTimeout(uint32_t ms) { return ms; }

    /// Alias class for asynchronous timeout control in milliseconds and
    /// repeats the test case handler with calling teardown and setup handlers
    inline control_t CaseRepeatAllOnTimeout(uint32_t ms) { return control_t(REPEAT_ALL_ON_TIMEOUT, ms); }
    /// Alias class for asynchronous timeout control in milliseconds and
    /// repeats only the test case handler without calling teardown and setup handlers
    inline control_t CaseRepeatHandlerOnTimeout(uint32_t ms) { return control_t(REPEAT_HANDLER_ON_TIMEOUT, ms); }

    class Case; // forward declaration

    /** Test setup handler.
     *
     * This handler is called before execution of any test case and
     * allows you to initialize your common test environment.
     *
     * @param   cases           the test cases in this test specification
     * @param   number_of_cases the total number of test cases in the test specification
     *
     * @returns
     *    You can return `STATUS_ABORT` if you initialization failed and the test teardown handler will
     *    then be called with the `REASON_SETUP`.
     */
    typedef status_t (*test_setup_handler_t)(const Case *const cases, const size_t number_of_cases);

    /** Test teardown handler.
     *
     * This handler is called after execution of all test case or if test execution is aborted.
     * You can use this handler to de-initialize your test environment and output test statistics.
     * The failure argument contains the immediate reason why this handler is called.
     * If the test completed normally without failures, this will contain `REASON_NONE`.
     *
     * After execution of this handler, the test harness will stop execution.
     *
     * @param   passed  the number of cases without failures
     * @param   failed  the number of cases with at least one failure
     * @param   failure the reason why this handler was called
     */
    typedef void (*test_teardown_handler_t)(const size_t passed, const size_t failed, const failure_t failure);

    /** Test failure handler.
     *
     * This handler is called anytime a failure occurs during the execution of a test speficication.
     * The handler only allows logging of failures and cannot influence test execution.
     *
     * @param   failure the reason why this handler was called
     */
    typedef void (*test_failure_handler_t)(const failure_t reason);

    /** Test case setup handler.
     *
     * This handler is called before execution of each test case and
     * allows you to modify your environment before each test case.
     *
     * @param   source          the test case to be setup
     * @param   index_of_case   the current index of the test case within the specification
     *
     * @returns
     *    You can return `STATUS_ABORT` to indicate that your setup failed, which will call the case
     *    failure handler with `REASON_SETUP` and then the case teardown handler with `REASON_SETUP`.
     *    This gives the teardown handler a chance to clean up a failed setup.
     */
    typedef status_t (*case_setup_handler_t)(const Case *const source, const size_t index_of_case);

    /** Primitive test case handler
     *
     * This handler is called only if the case setup succeeded and is followed by the test case teardown handler.
     *
     * @note This handler is executed only once.
     */
    typedef void (*case_handler_t)(void);

    /** Complex test case handler
     *
     * This handler is called only if the case setup succeeded and then may be repeated or
     * awaiting a asynchronous callback, depending on the return modifiers.
     *
     * @returns
     *    A combination of control modifiers.
     */
    typedef control_t (*case_control_handler_t)(void);

    /** Test case handler (repeatable)
     *
     * This handler is called only if the case setup succeeded and then may be repeated or
     * awaiting a asynchronous callback, depending on the return modifiers.
     *
     * @param   call_count    starting at `1`, contains the number of times this handler has been called
     *
     * @returns
     *    A combination of control modifiers.
     */
    typedef control_t (*case_call_count_handler_t)(const size_t call_count);

    /** Test case teardown handler.
     *
     * This handler is called after execution of each test case or all repeated test cases and
     * allows you to reset your environment after each test case.
     *
     * @param   source  the test case to be torn down
     * @param   passed  the number of cases without failures (can be >1 for repeated test cases)
     * @param   failed  the number failures (can be larger than the number of (repeated) test cases)
     * @param   failure the reason why this handler was called
     *
     * @returns
     *    You can return `STATUS_ABORT` to indicate that your teardown failed, which will call the case
     *    failure handler with `REASON_TEARDOWN`.
     */
    typedef status_t (*case_teardown_handler_t)(const Case *const source, const size_t passed, const size_t failed, const failure_t reason);

    /** Test case failure handler.
     *
     * This handler is called whenever a failure occurred during the setup, execution or teardown.
     *
     * @param   source  the test case in which the failure occurred
     * @param   reason the reason why this handler was called
     *
     * @returns
     *    You can return `STATUS_ABORT` to indicate that this failure is non-recoverable, which will call the case
     *    teardown handler with reason. If a failure occurs during teardown, the teardown will not be called again.
     *    You may return `STATUS_IGNORE` which will cause the harness to ignore and not count the failure.
     */
    typedef status_t (*case_failure_handler_t)(const Case *const source, const failure_t reason);


    // deprecations
    __deprecated_message("Use CaseRepeatAll instead.")     const control_t CaseRepeat            = CaseRepeatAll;
    __deprecated_message("Use CaseRepeatHandler instead.") const control_t CaseRepeatHandlerOnly = CaseRepeatHandler;

    __deprecated_message("Use REASON_NONE instead.")          const failure_reason_t FAILURE_NONE       = REASON_NONE;
    __deprecated_message("Use REASON_UNKNOWN instead.")       const failure_reason_t FAILURE_UNKNOWN    = REASON_UNKNOWN;
    __deprecated_message("Use REASON_CASES instead.")         const failure_reason_t FAILURE_CASES      = REASON_CASES;
    __deprecated_message("Use REASON_EMPTY_CASE instead.")    const failure_reason_t FAILURE_EMPTY_CASE = REASON_EMPTY_CASE;
    __deprecated_message("Use REASON_TIMEOUT instead.")       const failure_reason_t FAILURE_TIMEOUT    = REASON_TIMEOUT;
    __deprecated_message("Use REASON_ASSERTION instead.")     const failure_reason_t FAILURE_ASSERTION  = REASON_ASSERTION;
    __deprecated_message("Use REASON_CASE_SETUP instead.")    const failure_reason_t FAILURE_SETUP      = REASON_CASE_SETUP;
    __deprecated_message("Use REASON_CASE_TEARDOWN instead.") const failure_reason_t FAILURE_TEARDOWN   = REASON_CASE_TEARDOWN;
    __deprecated_message("Use REASON_IGNORE instead.")        const failure_reason_t FAILURE_IGNORE     = REASON_IGNORE;


}   // namespace v1
}   // namespace utest

#endif // UTEST_TYPES_H
